### DISCUSSION TITLE
using 3D velocity models in SW4 Green's function computation

### DISCUSSION THREAD

**Discussion by DanBrox** [geodynamics/sw4#156] [Q&A] [answered]
Possible to provide an example SW4 input file for computing Green's functions in a a non-trivial 3D velocity model?  I've got an example using a 1D (i.e. layered) velocity model which works (attached .in  as .txt:
[seismic.txt](https://github.com/geodynamics/sw4/files/10348039/seismic.txt)), but I'm not able to follow the manual instructions on using 3D velocity models,

**Comment by andersp** [comment]
@DanBrox we use 3-D material models all the time. Can you explain your issues with 3D velocity models? BTW, which version of sw4 are you using? I recommend the 3.0-beta2 release.

**Comment by DanBrox** [comment]
Trying to compute Green's functions in a 3D velocity model of southern Vancouver island to do moment tensor inversion with ISOLA.  1D velocity models are simple enough to define for Green's function computation, as in the seismic.txt file I attached, but I'm not sure which method from the SW4 user guide (https://geodynamics.org/resources/1767/download/SW4-UsersGuide-2.0.pdf) should be used to compute Green's functions in 3D velocity models, or how to implement the appropriate method in an SW4 input file.  'SW4, version 2.0' appears in the download directory.

**Comment by andersp** [comment]
First step is to install version 3 of sw4. It is available as a release under the "Code" tab on this page. Then you need to generate a 3D material model.  Depending on what data you have available this can be a major undertaking. Sw4 supports these file formats: pfile, ifile, sfile, and most recently gmg. Details of the formats are described in the user's guide.

**Reply by DanBrox** [reply]
Planning to download version 3 as you suggest.  What's the program for creating a 3D material model?  The output from this material model program is pfile, ifile, sfile, or gmg?  Which file structure would you suggest outputting for a velocity model which specifies elastic velocities at > 10000 grid points?  If there is an example of how this file should be interfaced to the SW4 input file this would be helpful - will try looking at the manual sections on this again but wasn't able to follow them the first few times.

**Comment by andersp** [comment, marked as answer]
If your data points are ordered as depth profiles on a regular grid in the horizontal directions, I recommend the pfile (for profile data) format.

**Reply by DanBrox** [reply]
Ok.  Thanks.  Should've read the user guide more closely for pfile format instructions.  Attached an example 3D velocity model SW4 input for southern Vancouver Island based on data from this reference (rename to *.ppmod): 

[vm.txt](https://github.com/geodynamics/sw4/files/10461680/vm.txt)
Savard G. Seismic velocity structure under Vancouver Island from travel time inversion: insight from Low Frequency Earthquakes (Doctoral dissertation, University of British Columbia).

**Comment by DanBrox** [comment]
Thanks.  I posted a solution to this on the discussion board.

On a different note, is it possible to simulate seismic wave propagation 
on a 2D grid in SW4, assuming elastodynamics variables (e.g. stress / 
strain) are independent of one of the spatial coordinates?  I know this 
could be done on a 3D grid in SW4, but I'd like to use the spatial 
symmetry of the solution to reduce the time required to compute 
solutions (this inquiry has also been posted on the discussion board).

**Reply by Astell-Chen** [reply]
Hi! Have you checked whether the calculated Green's function can inverse a correct moment tensor using the synthetic seismic waves which are calculated by SW4? In a homogenous model, I can get a correct result. But in a complex model (such as a layered model), I got a bad result.

**Reply by DanBrox** [reply]
Here are some notes I wrote on this topic available for download from my ResearchGate profile (https://www.researchgate.net/publication/368690450_Moment_Tensor_Inversion_of_Vancouver_Island_Earthquake_Data_Notes).  They're not peer reviewed/published since they don't contain original material, so if you see mistakes in them please let me know.  In the notes, there is an example of computing Green's functions with SW4 in a 3D layered model, and doing moment tensor inversion with ISOLA.  The resulting moment tensor parameters are similar to those obtained when Green's functions are computed with ISOLA forward simulation.
[MomentTensorInversion.pdf](https://github.com/geodynamics/sw4/files/10807501/MomentTensorInversion.pdf)

**Reply by Astell-Chen** [reply]
Thank you! I will read your manuscript carefully, and do some tests.
